### DISCUSSION TITLE
Difference between APCA Lc and DeltaE2000

### DISCUSSION THREAD

**Discussion by TaigaYamada** [Myndex/SAPC-APCA#108] [Q&A] [answered]
As I understand, APCA Lc is a perceptually uniform measure of lightness contrast.

In terms of perceptually uniform measure of lightness contrast, another popular measure is DeltaE2000.
According to the documentation on color.js by Lea Verou, DeltaE2000 is the "most accurate and most complicated Lab-based DeltaE algorithm to date"
https://colorjs.io/docs/color-difference#delta-e-e

Recently, there is another measure being discussed, the DeltaE based on OKLab. As I understand, this is just a euclidean distance in the OKLab color space.
https://www.w3.org/Graphics/Color/Workshop/slides/talk/lilley

When I was testing things out with APCA, I noticed that compared to DeltaE2000 or DeltaEOKLab, APCA Lc tended to become smaller in the comparison between darker colors than in comparison between lighter colors. I was wondering why this may be.

It is clear that APCA Lc and DeltaE formulas would never fully align. DeltaE formulas do not account for polarity, because there is no concept of foreground vs background. So my guess is that difference between APCA Lc and DeltaE arise from these differences designed specifically around text readability.

However, I don't know for certain, so I'd love to know more about the difference between APCA and DeltaE2000.

**Comment by Myndex** [comment]
Hi @TaigaYamada 

I’m not in the lab but here’s a short answer first:

deltaE2000 may be the “most complicated LAB based” color difference method…

…but…

That’s still leaving much to be desired. L\*a\*b\* is not a full color appearance model, and is limited. L\* is based on Munsell value, which is itself derived from experiments using diffuse reflecting paint tiles in a very controlled environment. And:

- Color is not real, it is a context sensitive perception. L\*a\*b\* Does very little for context.

- Since de2k, there is CAM02, CAM16 and many others, much more advanced.

- L*a*b* based difference measures are okay at finding very small differences—differences near the “JND” (just noticeable difference). And this is useful in industry, for instance for matching paint for a car, or paper quality, etc.

### None of that is supra threshold contrast.

The contrasts we are interested in for things like text for reading, are far away from the JND visibility threshold.

And the human vision system (HVS) has significantly different response characteristics Supra threshold for high spatial frequency stimuli (text).

That accounts for some of the differences.

The other factor: those models are often based on studies where the stimuli is a diffuse reflecting patch of very low spatial frequency.

But reading text on a self illuminated display is none of that, it’s very high spatial frequency and the light is being emitted/beamed straight into your eyes, often polarized, and with a gamma curve and things like anti-aliasing all in play.

And for higher spatial frequencies, it is the spatial characteristics (weight or line thickness) that drives contrast, **not** color (above a critical level).

That’s the simple overview, the minutia and nuance goes very deep here. Please feel free to follow up.

**Reply by TaigaYamada** [reply]
Thank you very much for the explanation!

So as I understand from your response, the differences arise from the following points:

| Point of difference | DeltaE | APCA |
|---|---|---|
| Target surface | Diffuse reflecting patches of low spatial frequency | On-screen text, which is high spacial frequency patterns on a self-illuminating surface |
| Focused level of contrast | Good for Just Noticeable Difference | Focused on readability contrast threshold, which is far higher than JND threshold |
| Consideration of context | Does not consider the context | Considers context like polarity, spatial characteristics |
| Advancement of the models | Models like CAM02, CAM16 are newer and more advanced than DeltaE2000 | APCA is one of the newest models to be developed |

Would you say the above points also apply to DeltaE based on OKLab?
It seems OKLab was specifically derived to perform similarly to CAM16 in terms of lightness and chroma: https://bottosson.github.io/posts/oklab/#how-oklab-was-derived

**Comment by Myndex** [comment]
Hi @TaigaYamada 

**Table row 1:** Yes

**Table row 2:** `L*a*b*` is really only good for small differences. CIEDE2000 (aka ∆E2000) was designed to improve that, and improve tracking over larger color distances, to be more accurate, I'd say "very small differences" instead of JND. 

**Table row 3:** ∆E2000 has some consideration of the context of the color it is being tested against. And `L*a*b*` has some assumptions relating to the defined test environment. The `L*` is based on Munsell value, and the Munsell system defines the context of a specific viewing environment. So, `L*a*b*` and therefore ∆E2000 at least specify a context.

**Table row 4:** While CIELAB is not quite a color appearance model, it is nevertheless one of the early attempts at perceptual uniformity in a color space. ∆E2000 isn't a model by itself, but a difference calculation for use with CIELAB. Those other models have their own color difference methods, namely euclidian distance if they are uniform.


### "What's a few deltas between colors..."

You might find this paper interesting, as it speaks to sample data and CIEDE2000: [ciede2000noteCRNA.pdf](https://hajim.rochester.edu/ece/sites/gsharma/ciede2000/ciede2000noteCRNA.pdf)

APCA is a _contrast_ appearance model, and by itself not a color appearance model. SACAM (in development) is the larger project that APCA is a part of. (SACAM is "Sjuv Accessible Color Appearance Model"). ∆E2000 is very weighted for hue/chroma. APCA is much more achromatic, as achromatic luminance is what is required for readability.

Some examples in the next post showing when the ∆E value gets higher, the meaningfullness of that value is questionable. And also **how the ∆E2000 value does not relate to readable contrast.**


> ..._Would you say the above points also apply to DeltaE based on OKLab?_...

Hmmm... ∆E2000 was designed to work with _CIELAB_ specifically, which has some issues as we know. OKLab was designed to fix some of those issues itself. Thus, I'm not sure how an unchanged ∆E2000 will work with OKLab which is different on a number of levels, including the actual values. At the very least, the constants used with ∆E2000 would need to be adjusted.

> ..._It seems OKLab was specifically derived to perform similarly to CAM16 in terms of lightness and chroma_...

It's interesting that Bjorn used CAM16 to generate the training data... that partly explains some questions I've had. I know of some who are using OKLab with APCA replacing the L. 


## Discussion of contrast vs color
But also: "color distance" is more of a measurement (insofar as color, which is only a perception, can be "measured"). Contrast is a perception, and _not_ a function of color distance. Here are some visual examples.


### Context
In this display, both the yellow dots are exactly the same color comming out of the display, as are the grey swuares they are on. Do they look the same?

![YellowDotCheckerShadow_DLyon_PD](https://github.com/Myndex/SAPC-APCA/assets/42009457/6c3665a8-7701-4345-aa10-c7c5316c4cdb)

The context of the shadow makes the yellow dot under the shadow appear brighter than the dot not in shadow. In the same image, and even close to each other. 


### Edge Detection
APCA is focused on the "perception of edge contrast" as opposed to the "perception of a given color". And this is a non-trivial difference. Take a look at this row of low spatial frequency gradients:

![mach bands](https://github.com/Myndex/SAPC-APCA/assets/42009457/14f4384c-c521-4583-accb-b4bc0c9fa247)

Except there are actually no gradients in this row whatsoever. Each grey patch is the same exact grey within its rectangle, but at every edge adjacent to the next patch, the human vision system "enhances" the edge, by ramping up or ramping down the grey leading up to the edge. 

The edge itself is a very high spatial frequency, connected to the low spatial of the rectrangle. The effect of the ramp up/down leading to the edge is called "Mach bands", and it is a perception entirely within the human vision system.


### Spatial Resolution
Here's an example of how spatial characteristics are the main driver of contrast. In the following graphic, all of the grey text samples are at the exact same color. Thus they all have the same color distance. Are they all equally readable?

![New contrast sensitivity Graph v5 new leads all same color](https://github.com/Myndex/SAPC-APCA/assets/42009457/d31ecd96-27ee-454e-b502-8253c952169f)

As spatial freq. rises above about 6 to 8 cycles per degree, contrast sensitivity plummets. Taking a 2px thick line and reducing in 50% to 1px results in a drop in contrast sensitivity by over an order of magnitude.

**Reply by Myndex** [reply, marked as answer]
## ANNEX • _∆E2000 vs APCA_
### ∆E2000 versus APCA Lc value (a low contrast comparison).
To cap all this off here are some examples of ∆E2000, which is heavilly weighted for chroma and hue, and comparisons are made to the APCA contrast Lc value which is largely achromatic, as achromatic luminance is what the Visual Word Form Area uses for letter pair/word detection.

In the examples below, the ∆E2000 values were from Bruce Linbloom's site, and the APCA examples are from the SAPC research tool which shows LC values less than 1, though APCA can round to nearest whole number, which I what I'll mention below.

Here's the center test target as used below, but at a high contrast of Lc 75 so you can see the details and the small thin text.

<img alt="Screen Shot 2023-10-10 at 4 36 45 AM" width="160" src="https://github.com/Myndex/SAPC-APCA/assets/42009457/73a7930a-13ba-4c96-a66a-5003e741a71d">

In the later low contrast examples, notice how you can not see the small text, and barely see the big bold ampersand. The small thin text is essentially invisible, even in the first high ∆E example.

Color as in hue/chroma play their biggest roll at very low spatial resolutions, at high spatial, luminance, not color, is dominant. APCA lightness contrast (Lc) is an achromatic luminance, weighted (powercurve) for perceptual lightness.  ∆E2000 is substantially weighted toward hue/chroma.


----
### ∆E 38
To start off, here is a DE 38, at Lc 0 

<img alt="De 38 Lc 0 2" width="720" src="https://github.com/Myndex/SAPC-APCA/assets/42009457/e7a0cfae-3d85-4c16-9f4f-5b679d113f43"><br>
This is the highest ∆E I'll be showing here.

_Calculation from BruceLinbloom.com_ :     
<img alt="De 38 Lc 0 2 c" width="170" src="https://github.com/Myndex/SAPC-APCA/assets/42009457/a818363c-f61b-40fe-89c6-b1fd475fb8e6">


### ∆E 16
Next, about half the above ∆E at ∆E 16, and again Lc 0     

<img alt="De 16 6 Lc 0" width="720" src="https://github.com/Myndex/SAPC-APCA/assets/42009457/14f22436-f5aa-4c81-9719-421d695cc5de"><br>
<img alt="De 16 6 Lc 0 c" width="170" src="https://github.com/Myndex/SAPC-APCA/assets/42009457/e7b25cac-9328-4fd7-a777-38ec9491f382">

### ∆E 8
Half the previous at ∆E 8, yet a slightly higher, Lc 3

<img alt="De 8 Lc -3" width="720" src="https://github.com/Myndex/SAPC-APCA/assets/42009457/7d5c21e8-e52e-453e-b8d8-6131e922e8a6"><br>
Oh but wait, it's half the ∆E yet a slightly **_more_** visible difference?

<img alt="De 8 Lc -3 c" width="170" src="https://github.com/Myndex/SAPC-APCA/assets/42009457/58511b49-ef15-4a87-a686-c8b63a819f6c">



### ∆E 21
Here is ∆E 21, and Lc 3

<img alt="De 21 Lc -3" width="720" src="https://github.com/Myndex/SAPC-APCA/assets/42009457/8c4a8a8a-ecef-4e02-b21e-fff4024631c7"><br>
<img alt="De 21 Lc -3 c" width="170" src="https://github.com/Myndex/SAPC-APCA/assets/42009457/7a98e518-f1a7-4c26-a463-aacc1dd42c75">

### ∆E 6
Compare the above to the next which is less than a third of the above at ∆E 6, but the Lc is higher at 10.

<img alt="De 6 Lc 10" width="720" src="https://github.com/Myndex/SAPC-APCA/assets/42009457/718fe9bb-59cd-4ebf-90de-0a2b86eb83d9"><br>
<img alt="De 6 Lc 10 c" width="170" src="https://github.com/Myndex/SAPC-APCA/assets/42009457/a8f8160d-4935-4cb0-9359-ef3662165ad0">


### ∆E 9
The final examples, here ∆E 9 and Lc 0    
<img alt="De 9 Lc 0 " width="720" src="https://github.com/Myndex/SAPC-APCA/assets/42009457/3d9be702-bd64-4c8c-a284-cb44c6fe5674"><br>
<img alt="De 9 Lc 0 c" width="170" src="https://github.com/Myndex/SAPC-APCA/assets/42009457/39b658a8-367f-4125-90a4-8687623a1002">

### ∆E 7
And this is a bit less at ∆E 7, but the Lc is higher at Lc 11.    
<img alt="De7 Lc11" width="720" src="https://github.com/Myndex/SAPC-APCA/assets/42009457/ad7627f1-2de3-479f-aac6-73b1a53a8daa"><br>
<img alt="De7 Lc11 c" width="170" src="https://github.com/Myndex/SAPC-APCA/assets/42009457/d78554a1-d515-4015-8e0e-04bd5610520f">


## Upshot

What I hope this demonstrates: to answer why ∆E2000 is not a useful measure for contrast, and also, how APCA Lc values are more in tune with predicting appearance of high spatial resolution elements, as needed for reading.

Regarding the low Lc values: In the APCA Readability Criterion, the low-end cut off for text is Lc 30. The very low Lc values above were only to show the sensitivity to lightness changes that is part of APCA, but not so much ∆E2000.

**Reply by TaigaYamada** [reply]
@Myndex thank you so much for such a detailed explanation! I understand much more about the APCA Lc and CIEDE2000 now.

Thank you for taking your time.

**Comment by Myndex** [comment]
## Revisit of the OP
Having posted the several recent posts, I re-read the OP, and I think I may have gotten something slightly off. 

### 1) OK ∆E
I misread initially, with the impression someone was applying CIEDE2000 (∆E2000) to OKlab, but on re-reading I see just euclidian distances, i.e.

$` \Delta E^*_{ab} = \sqrt{(L^*_2 - L^*_1)^2 + (a^*_2 - a^*_1)^2 + (b^*_2 - b^*_1)^2} \\  `$

This is a simple euclidian distance—the straight line between two colors. But as discussed above, this is not contrast for the purpose of readability.

The simple euclidian distance is only good for a perceptually uniform cartisian color space. And then we might ask, perceptually uniform on which axes? As I implied in some of what I wrote above, there are really more than three axes defining color appearance.

### 2) APCA ∆E...
> ..._I noticed that compared to DeltaE2000 or DeltaEOKLab, APCA Lc tended to become smaller in the comparison between darker colors than in comparison between lighter colors._..

1) Simplicity of use (only a pair of colors tested)
2) Common use case of a bright environment with a self illuminated display
3) Center of contrast is higher than middle grey, especially in a brighter surround
4) WCAG 2, L*, and other measures based on diffuse reflecting paint patches, and with a middle grey of 18%, may be aligned with overall lightness or color in the context of low spatial grdients, but are not aligned with high spatial resolution related visual contrasts.


It takes more than two colors in most cases to determine the expected contrast. More than would be convenient for a designer's guidelines, or for a public standard. The WCAG 2.x tests onlhy a pair, so the initial APCA is set for just a pair, with some reasonable assumptions built in regarding expected use cases, and the optimum repsonse for those use cases.

A bright environment is the most challenging for reading on a self-illuminated display (e-ink is a different story, as of course, is traditional print). And in a bright environment, dark color pairs become increasingly unreadable.

And in our empirical studies, we found the "worst expected common case" is a bright office or in shade on a sunny day. And here, dark colors are negatively affected. But light colors are not negatively affected to such a degree in darker environments.

L*, WCAG 2.x, and several others, end up over reporting contrast for dark color pairs, and in a way that results in unreadable values, especially for WCAG 2x contrast math.

### _Solution_
To accomodate the "general worst common case" of use for devices and displays, keep things to a simple pair of colors, align to the center of contrast on a display, and so forth, the curves are optimized for the expected environments. As such, in comparison to non-contrast oriented models, APCA does not over-report dark color pairs.

### 3) Color Chiropractor?
> _It is clear that APCA Lc and DeltaE formulas would never fully align._ 

I'd suggest they will never align at all — they are both designed for and serving very different purposes.


Please don't hesitate to follow up if you have additional questions.
